ui/components/app/snaps/snap-permissions-list/snap-permissions-list.js

```diff
@@ -1,52 +1,129 @@
 import { useI18nContext } from '../../../../hooks/useI18nContext';
-import { Box } from '../../../component-library';
+import { Box, ButtonLink } from '../../../component-library';
 import {
   getMultipleTargetsSubjectMetadata,
   getSnapsMetadata,
 } from '../../../../selectors';
+import {
+  Display,
+  FlexDirection,
+  JustifyContent,
+} from '../../../../helpers/constants/design-system';
+import SnapPermissionAdapter from '../snap-permission-adapter';
+import { PermissionWeightThreshold } from '../../../../../shared/constants/permissions';
 import { getSnapName } from '../../../../helpers/utils/util';
-import SnapPermissionCell from '../snap-permission-cell';
+import { getWeightedPermissions } from '../../../../helpers/utils/permission';
 
 export default function SnapPermissionsList({
   snapId,
   snapName,
   permissions,
   connections,
   showOptions,
+  turnOffAbstraction,
+  showAllPermissions,
 }) {
   const t = useI18nContext();
-  const snapsMetadata = useSelector(getSnapsMetadata);
-  const permissionsToShow = {
-    ...permissions,
-    connection_permission: connections ?? {},
-  };
+
+  const combinedPermissions = useMemo(() => {
+    return { ...permissions, connection_permission: connections ?? {} };
+  }, [permissions, connections]);
+
   const targetSubjectsMetadata = useSelector((state) =>
     getMultipleTargetsSubjectMetadata(state, connections),
   );
 
-  return (
-    <Box className="snap-permissions-list">
-      {getWeightedPermissions({
+  const snapsMetadata = useSelector(getSnapsMetadata);
+
+  const [showAll, setShowAll] = useState(false);
+
+  const [permissionsToDisplay, setPermissionsToDisplay] = useState([]);
+
+  const [weightedPermissions, setWeightedPermissions] = useState({});
+
+  useEffect(() => {
+    let finalPermissions = weightedPermissions;
+    if (Object.keys(finalPermissions).length === 0) {
+      finalPermissions = getWeightedPermissions({
         t,
-        permissions: permissionsToShow,
+        permissions: combinedPermissions,
         subjectName: snapName,
         getSubjectName: getSnapName(snapsMetadata),
-      }).map((permission, index) => (
-        <SnapPermissionCell
+      });
+      setWeightedPermissions(finalPermissions);
+    }
+
+    if (turnOffAbstraction && !showAll) {
+      setShowAll(true);
+    }
+
+    if (Object.keys(finalPermissions).length <= 3 && !showAll) {
+      setShowAll(true);
+    }
+
+    if (showAll) {
+      setPermissionsToDisplay(
+        <SnapPermissionAdapter
+          permissions={finalPermissions}
           snapId={snapId}
+          snapName={snapName}
+          targetSubjectsMetadata={targetSubjectsMetadata}
           showOptions={showOptions}
-          connectionSubjectMetadata={
-            targetSubjectsMetadata[permission.connection]
-          }
-          permission={permission}
-          index={index}
-          key={`permissionCellDisplay_${snapId}_${index}`}
-        />
-      ))}
+          weightThreshold={Infinity}
```

Comment:
```suggestion
```

You can omit this since you're using `weightThreshold ?? Infinity` in the component.